### PR TITLE
Feature/namelist

### PR DESCRIPTION
**Pull request by mickaelaccensi**
# Pull Request Summary
updates to be able to run all the regtests with namelist input files

## Description
this PR aims in having WW3 running all the regtests with namelist input files and having the same results as the version with inp files. To obtain this, many improvements, updates and bugfixes have been performed.

### Issue(s) addressed

- [X] ww3_bound with regtest ww3_tr1
- [ ] set -N option for all matrix_* to enable namelist use by default
- [X] set -o both option for all matrix_* to enable netcdf and ascii outputs by default
- [X] fix issue #421
- [ ] ww3_tp2.17 run correctly now for ww3_grib since new library installation
- [ ] ww3_tp2.15  works well with namelist
- [ ] update ww3_multi.nml and ww3_shel.nml with atmos forcing in regtests
- [ ] create missing ww3_multi.nml and ww3_shel.nml in a few regtests
- [X] reorganize ww3_tp2.14 by input subdirectories to avoid files renaming inside input dir (different cases if inp or nml)
- [X] update ww3_from_ftp.sh and matrix.base to have ww3_tp2.14 working with namelist (issue #513 )
- [ ] remove wrong test to disable restart2 per grid to avoid crash on regtest ww3_tp2.21

### Commit Message
updates to be able to run all the regtests with namelist input files

### Check list  

- [X] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [X] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [X] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [X] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested?
first matrix with inp files and comparisons between develop and namelist branch
then matrix nml files and comparisons between develop and namelist branch
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
yes by existing regtests
* Have the matrix regression tests been run (if yes, please note HPC and compiler)?
datarmor with intel compiler
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):    

**for matrix with inp files :** 
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7987311/matrixDiff.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7987314/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7987315/matrixCompFull.txt)

**for matrix nml files :**
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7988072/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7988073/matrixCompSummary.txt)
[matrixDiff.zip](https://github.com/NOAA-EMC/WW3/files/7988079/matrixDiff.zip)

* Please indicate the expected changes in the regression test output ([Note the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).

**for matrix with inp files :** 
-mww3_test_03 : known not b4b
-ww3_tnc1 : forced to use namelist in regtest - updated dates compared to ww3_shel.inp
-ww3_tp2.14 : reorganize but separated input directories 

**for matrix with nml files :** 
mww3_test_02 : ww3_multi.nml.log (output of missing DOMAIN namelist)

mww3_test_03 : ww3_multi.nml.log (output of missing DOMAIN namelist)
                                work_PR1_MPI_d2/ww3.196806.nc_diff.txt (usual not b4b difference)
                                work_PR2_UNO_MPI_d2/ww3.196806.nc_diff.txt (usual not b4b difference)
                                work_PR2_UQ_MPI_d2/ww3.196806.nc_diff.txt (usual not b4b difference)
                                work_PR2_UQ_MPI_d2/ww3.196806_spec.nc_diff.txt (usual not b4b difference)
                                work_PR3_UNO_MPI_d2_c/ww3.196806.nc_diff.txt (usual not b4b difference)
                                work_PR3_UNO_MPI_d2_c/ww3.196806_spec.nc_diff.txt (usual not b4b difference)
                                work_PR3_UNO_MPI_d2/ww3.196806.nc_diff.txt (usual not b4b difference)
                                work_PR3_UNO_MPI_d2/ww3.196806_spec.nc_diff.txt (usual not b4b difference)
                                work_PR3_UNO_MPI_d2/ww3.68060608.hs_diff.txt (usual not b4b difference)
                                work_PR3_UQ_MPI_d2_c/ww3.196806.nc_diff.txt (usual not b4b difference)
                                work_PR3_UQ_MPI_d2_c/ww3.196806_spec.nc_diff.txt (usual not b4b difference)
                                work_PR3_UQ_MPI_d2_c/ww3.68060605.hs_diff.txt (usual not b4b difference)
                                work_PR3_UQ_MPI_d2/ww3.196806.nc_diff.txt (usual not b4b difference)
                                work_PR3_UQ_MPI_d2/ww3.196806_spec.nc_diff.txt (usual not b4b difference)
                                work_PR3_UQ_MPI_d2/ww3.68060605.hs_diff.txt (usual not b4b difference)

mww3_test_04 : ww3_multi.nml.log (output of missing DOMAIN namelist)

mww3_test_05 : ww3_multi.nml.log (output of missing DOMAIN namelist)

mww3_test_06 : ww3_multi.nml.log (output of missing DOMAIN namelist)

mww3_test_07 : ww3_multi.nml.log (output of missing DOMAIN namelist)

mww3_test_08 : ww3_multi.nml.log (output of missing DOMAIN namelist)

mww3_test_09 : ww3_multi.nml.log (output of missing DOMAIN namelist)
                           work_MPI/log.Huron_diff.txt  (X instead of L)
                           work_MPI/log.Michi_diff.txt  (X instead of L)
                           work_MPI/log.Super_diff.txt  (X instead of L)

ww3_tic1.1 : work_IC5_M2/ww3_shel.nml.log_diff.txt (input path corrected)

ww3_tnc1 : work/ww3_shel.nml.log_diff.txt (output dates updated)

ww3_tp2.14 :      work_OASACM2/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASACM2/ww3_grid.out_diff.txt (input path updated)
                           work_OASACM2/ww3_ounf.out_diff.txt (remove ICE IC1 IC5 TWI not requested in ww3_shel)
                           work_OASACM3/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASACM3/ww3_grid.out_diff.txt (input path updated)
                           work_OASACM3/ww3_ounf.out_diff.txt (remove ICE IC1 IC5 TWI not requested in ww3_shel)
                           work_OASACM3/ww3_shel.nml.log_diff.txt (values updated with inp version)
                           work_OASACM4/log.ww3_diff.txt (use restart from oasacm3)
                           work_OASACM4/OUTPUT_TOY.txt_diff.txt (due to restart file)
                           work_OASACM4/TOY__CHA_toyexe_03.nc_diff.txt (due to restart file)
                           work_OASACM4/ww3.201109.nc_diff.txt (corrected values)
                           work_OASACM4/WW3_ACHA_wwatch_03.nc_diff.txt (due to restart file)
                           work_OASACM4/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASACM4/ww3_grid.out_diff.txt (input path updated)
                           work_OASACM4/ww3_ounf.out_diff.txt (remove ICE IC1 IC5 TWI not requested in ww3_shel)
                           work_OASACM5/OUTPUT_TOY.txt_diff.txt (10-6 differences)
                           work_OASACM5/TOY__CHA_toyexe_03.nc_diff.txt (corrected values)
                           work_OASACM5/ww3.201109.nc_diff.txt (corrected values)
                           work_OASACM5/WW3_ACHA_wwatch_03.nc_diff.txt (10-6 differences)
                           work_OASACM5/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASACM5/ww3_grid.out_diff.txt (input path updated)
                           work_OASACM5/ww3_ounf.out_diff.txt (remove ICE IC1 IC5 TWI not requested in ww3_shel)
                           work_OASACM6/namcouple_diff.txt (input path updated)
                           work_OASACM6/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASACM6/ww3_ounf.out_diff.txt (remove ICE IC1 IC5 TWI not requested in ww3_shel)
                           work_OASACM/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASACM/ww3_grid.out_diff.txt (input path updated)
                           work_OASACM/ww3_ounf.out_diff.txt (remove ICE IC1 IC5 TWI not requested in ww3_shel)
                           work_OASICM/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASICM/ww3_grid.out_diff.txt (input path updated)
                           work_OASOCM/ww3_grid.nml.log_diff.txt (input path updated)
                           work_OASOCM/ww3_grid.out_diff.txt (input path updated)
                           work_OASOCM/ww3_ounf.out_diff.txt (remove ICE IC1 IC5 TWI not requested in ww3_shel)

ww3_tp2.15 : work_PR3_UQ_RHO_MPI/ww3_shel.nml.log_diff.txt (input path corrected)
                      work_PR3_UQ_RHO/ww3_shel.nml.log_diff.txt (input path corrected)
                      work_ST4FLX5/ww3_shel.nml.log_diff.txt (input path corrected)
                      work_ST6FLX5/ww3_shel.nml.log_diff.txt (input path corrected)

ww3_tp2.17 : work_a/ww3_shel.nml.log_diff.txt (corrected output field/point/restart start date)
                      work_b/ww3_shel.nml.log_diff.txt (corrected output field/point/restart start date)
                      work_c/ww3_shel.nml.log_diff.txt (corrected output field/point/restart start date)
                      work_ma/ww3_multi.nml.log_diff.txt (corrected output field/point/restart start date)
                      work_mb/ww3_multi.nml.log_diff.txt (corrected output field/point/restart start date)
                      work_mc/ww3_multi.nml.log_diff.txt (corrected output field/point/restart start date)

ww3_tp2.21 : work_ma/log.mww3_diff.txt(space)
                      work_mb/log.mww3_diff.txt (space)

ww3_ts1 : work_NL5/ww3_shel.nml.log_diff.txt (input path corrected)

ww3_ts4 : ww3_multi.nml.log (output of missing DOMAIN namelist)

ww3_ts3 : ww3_shel.nml (corrected input current forcing fields as homogeneous instead of disabled)
                 all ouputs are differents due to homogeneous current input forcing

### CODE REVIEW

File: model/src/w3nmlboundmd.F90 [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+!/ ------------------------------------------------------------------- /
+
+
+
+
+END MODULE W3NMLBOUNDMD
+
+!/ ------------------------------------------------------------------- /
+
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3nmlboundmd.F90:342]: delete the many blank lines at the end of this routine?

File: model/src/wminitmd.F90 [hunk 2]
```diff
@@ -5607,13 +5603,6 @@ SUBROUTINE WMINITNML ( IDSI, IDSO, IDSS, IDST, IDSE, IFNAME,       &
 !
 ! Checkpoint
         J=8
```
Comment (JessicaMeixner-NOAA) [inline model/src/wminitmd.F90:5605]: Why was this deleted?

File: regtests/bin/matrix_ncep [hunk 3]
```diff
@@ -164,9 +164,9 @@
 # Compile option
   if [ "$cmplOption" = 'y' ]
   then
-     opt="-c $cmplr -S -T"
+     opt="-o both -c $cmplr -S -T -N"
```
Comment (JessicaMeixner-NOAA) [inline regtests/bin/matrix_ncep]: Please revert this change.  We will not run with the -N option until all programs are available to be run with namelists, until then we'll continue the regression testing using inp files.

File: regtests/bin/matrix_milhydro [hunk 4]
```diff
@@ -65,9 +65,9 @@
 # Compile option
   if [ "$cmplOption" = 'y' ]
   then
-     opt="-c $cmplr -S -T"
+     opt="-o both -c $cmplr -S -T -N"
```
Comment (JessicaMeixner-NOAA) [inline regtests/bin/matrix_milhydro]: @thesser1 @maryanderson64 please indicate if you are okay with this change

File: regtests/bin/matrix_s4 [hunk 5]
```diff
@@ -121,9 +121,9 @@
 # Compile option
   if [ "$cmplOption" = 'y' ]
   then
-     opt="-c $cmplr -S -T"
+     opt="-o both -c $cmplr -S -T -N"
```
Comment (JessicaMeixner-NOAA) [inline regtests/bin/matrix_s4]: Please revert this change (same comment as for the NCEP above).

File: regtests/bin/matrix_ukmo_cray [hunk 6]
```diff
@@ -104,9 +104,9 @@ fi
 
   if [ "$cmplOption" = 'y' ]
   then
-     export rtst="./bin/run_test -c $cmplr -S"
+     export rtst="./bin/run_test -o both -c $cmplr -S -N"
```
Comment (JessicaMeixner-NOAA) [inline regtests/bin/matrix_ukmo_cray]: @ukmo-ccbunney please indicate if you are okay with this change

File: regtests/bin/matrix_ncep (hunk 3 above)
Comment (JessicaMeixner-NOAA) [reply regtests/bin/matrix_ncep]: Also please remove -o both.  This should be in the matrix.base not an option determined here.

File: regtests/mww3_test_05/input/ww3_multi_grdset_d.nml [hunk 7]
```diff
@@ -41,6 +41,8 @@
 !     INPUT(I)%FORCING%WATER_LEVELS  = F
 !     INPUT(I)%FORCING%CURRENTS      = F
 !     INPUT(I)%FORCING%WINDS         = F
+!     INPUT(I)%FORCING%ATM_MOMENTUM  = F
```
Comment (JessicaMeixner-NOAA) [inline regtests/mww3_test_05/input/ww3_multi_grdset_d.nml:44]: I thought we decided we were not going to update this type of thing in the regression tests?  There seem to be a lot of changes there are non-changes///changes in comments so it's hard to see what is a change and not a change.

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
A few comments: 

1. Please check with each of the groups that run the different matrix_* options about if they're okay with running the -N option.  At NCEP, the plan was to hold off on this until we have all available files using namelist.  The -o both option should be an update in the matrix.base, not as a blanket choice for every test. 
2. There are a lot of comment changes in the regression test files. I thought we were just going to update what is in the example input/namelist and stop changing this so much in the regression tests?  Not updating these files could help with not having 402 file changes in this PR. 
3. It's really hard to tell what's different development.  It seems like this is a few PRs worth of updates that might should be separated so we know what's going on?   For example (1) updating regtests for namelist option changes (2) adding the new ww3_bound namelist option (3) other bug fixes.  Thoughts?

File: regtests/bin/matrix_ukmo_cray (hunk 6 above)
Comment (ukmo-ccbunney) [reply regtests/bin/matrix_ukmo_cray]: @JessicaMeixner-NOAA @mickaelaccensi - Yes, that's fine with me. Thanks for checking.

Comment (ukmo-ccbunney) [reply regtests/bin/matrix_ukmo_cray]: Just to clarify, I am happy with the addition of the `-N` flag here and agnostic about the `-o both` flag (happy for it to be either here or in the matrix.base)

**Comment by ukmo-ccbunney** [comment]
@mickaelaccensi After saying that I was happy with the `-o both` flag in the `matrix_*` files, I note that this does cause a bit of an issue where the `-o` flag is set in the `matrix.base` already, for example at line 2252:
```
echo "$rtst -s ST6      -w work_ST6  -o netcdf  $ww3 ww3_ts1" >> matrix.body
```

Having `-o both` set in the `matrix_*` generation script (`matrix_ukmo_cray` in my case) results in the following entry in the generated `matrix` file:
```
./bin/run_test -o both -c ukmo_cray_gnu -S -N -s ST6      -w work_ST6  -o netcdf  ../model ww3_ts1
```
Note the two differing `-o both` and `-o netcdf` flags. I assume in this case `netcdf` will take precedence.

I think we will need to  decide whether we want a blanket `-o` flag set in the matrix_* files (I know @JessicaMeixner-NOAA doesn't) or at a per-test level in the `matrix.comp`, but I don't think we can have both! :)

**Comment by JessicaMeixner-NOAA** [comment]
To me, the $rtst are more system dependent instead of test dependent.    I'm not opposed to having everything have both, some tests don't have netcdf or ascii or vice-versa, however we do have an issue that we've added a few tests with grib output and so @aliabdolali created a -b all option for that.    So we need to have some consistency.  Regardless, I would say this is out of the scope of making namelist options work and therefore should not be in this PR and should be done by itself as that's changing a lot of tests and we should have that be independent of other commits. 


We also have a few tests that are basically the same except one has the -N option.  So when we do finally have all programs using namelist and that becomes the official over the inp then, we could remove those duplicative tests.

File: model/bin/ww3_from_ftp.sh [hunk 8]
```diff
@@ -10,19 +10,32 @@ curr_dir=`pwd`
...
 ww3ver=v7.12.6
 
+interactive='y'
+if [ $# -eq 3 ] ; then
+  interactive='n'
+  ww3dir=$1  # [ '../../']
+  wnew=$2    # ['y']
+  wnew2=$3   # ['y']
+fi
+
```
Comment (aliabdolali) [inline model/bin/ww3_from_ftp.sh]: Does it solve https://github.com/NOAA-EMC/WW3/issues/578? if yes, it should be a separate PR.

**Comment by JessicaMeixner-NOAA** [comment]
@mickaelaccensi I think there are still some separate parts that could be pulled out of this.  We've already pulled 2 parts (ww3_from_ftp.sh) and the (-b both/all).  A few things I could see as maybe being separate are: 
-- simplification of ww3-tp2.14 regtests 
-- the namelist file regtests updates (maybe?) 
-- ???? 
Again my computer won't even load the diffs because there are so many so you probably know better than I do.  Either way we need to merge in the other two changes in here.  Let me know when you're ready for me to re-review this.

**Comment by mickaelaccensi** [comment]
> @mickaelaccensi I think there are still some separate parts that could be pulled out of this. We've already pulled 2 parts (ww3_from_ftp.sh) and the (-b both/all). A few things I could see as maybe being separate are: -- simplification of ww3-tp2.14 regtests -- the namelist file regtests updates (maybe?) -- ???? Again my computer won't even load the diffs because there are so many so you probably know better than I do. Either way we need to merge in the other two changes in here. Let me know when you're ready for me to re-review this.

yes, I'm currently working on tp2.14 branch which should be ready today
For the namelist files in regtests, I was thinking in cleaning up the nml files as discussed in #572

**Comment by JessicaMeixner-NOAA** [comment]
@mickaelaccensi I know that this PR was broken into several PRs, but I'm not sure if all of it has been merged in or if this combined PR can be closed.  Thoughts?

**Comment by mickaelaccensi** [comment]
it can be closed
